### ISSUE TITLE
5.0: Scrolling in dashboard is sluggish

### ISSUE THREAD

**Issue by dictvm** [grafana/grafana#11053]
We have just upgraded our Grafana deployment to 5.0. It's awesome, thanks for that. However, we noticed a performance regression when scrolling inside of Grafana. It feels janky/sluggish, at least on macOS.

- What Grafana version are you using? 5.0
- What datasource are you using? InfluxDB, Postgres, Prometheus (but it doesn't matter for this issue)
- What OS are you running grafana on? macOS High Sierra 10.13.3 (17D102).
- What did you do? I'm scrolling in a dashboard.
- What was the expected result? Scrolling should feel natural/native.
- What happened instead? Scrolling feels janky/sluggish, as if the browser's scrolling behaviour is hijacked via Javascript or similar means.
- If related to metric query / data viz: n/a
  - Include raw network request & response: n/a

**Comment by torkelo** [comment]
what browser are you using?

**Comment by torkelo** [comment]
In chrome and firefox on OSX I can't notice any sluggishness , we are using a js framework to handle the scrollbar as the native scrollbar takes up space and looks absolutely horrible (and not all browser allow you to change style (nor overlay them)

**Comment by dictvm** [comment]
Oh, sorry, in a previous version of the text I mentioned that I tested this with Chrome 64 64.0.3282.186 (Official Build) (64-bit) and Safari Version 11.0.3 (13604.5.6).

**Comment by tlex** [comment]
I can confirm, if you have a dashboard with many panels it moves slower than on the latest 4.x. Tested on both Chrome 64.0.3282.186 (Official Build) (64-bit) - MacOS and on Firefox on Android (same dashboard)

**Comment by thenayr** [comment]
Chiming in here, noticeable difference in scrolling between 5.0 and earlier versions.  Feels much choppier.

**Comment by antiquechrono** [comment]
Same Issue, It is also happening on IOS. Grafana 4.x scrolls very smoothly.

macOS Sierra - 10.12.6

Desktop Safari Version 10.1.2 (12603.3.8)
IOS 10.3.2 - Safari 10
Chrome Version 64.0.3282.186 (Official Build) (64-bit)

**Comment by torkelo** [comment]
I am also seeing it now, after a few seconds of scrolling it happens

**Comment by daniellee** [comment]
Trying to recreate this on mac and linux here but not seeing what happens on Torkel's mac. When Torkel showed this to us on Saturday, then scrolling stopped working after 2 or 3 seconds. Not just slower but actually stopped working.

I can however recreate this problem that seems related where scrolling in panels on mobile does not work: https://community.grafana.com/t/grafana-5-overflow-scroll-doest-work-on-mobile/5480/5

And I have made a test dashboard for that:

http://play.grafana.org/d/JGi4Bmekz/scrolltest?orgId=1

@thenayr @tlex @dictvm it would be very helpful if one of you could make a video (like in the community issue linked to above) illustrating the difference between 4.6 and 5.0. I'm not sure I really see any difference. 

Three public sites to test 5.0 on are:

- http://play.grafana.org
- https://prometheus.devstats.cncf.io/d/UeP_tSqkz/commits-repository-groups?orgId=1
- https://k8s.devstats.cncf.io/d/UeP_tSqkz/commits-repository-groups?panelId=2&orgId=1

A public site to test an older version (4.3) on:

- https://monitor.gitlab.net/dashboard/db/git-timings?refresh=30s&orgId=1

**Comment by daniellee** [comment]
Ok, on iOS 11.2.6 I'm seeing that the scrolling is slower on large dashboards. Will test disabling the [perfect-scrollbar](https://github.com/utatti/perfect-scrollbar) for mobile and see if there is a major difference.

But still have not seen the scrolling hang so far.

**Comment by dictvm** [comment]
Alright, here are the two videos you requested. However, due to the dashboards not being very complex, it's rather difficult to see the difference. In the video showing the behavior of Grafana 5.0 I was scrolling smoothly using the trackpad but Grafana didn't scroll accordingly and ignored a few scrolling gestures. What is quite noticeable is the difference in general scrolling behavior: On 4.* Chrome is able to overscroll/bounce when you scroll to the end of a page, which is the native macOS behavior, while in 5.0 it just stops at the end of the page.

[Video 4](https://nofile.io/f/FD5HZZ0ZVvB/4.mov)
[Video 5](https://nofile.io/f/v3p3wbG0Qs4/5.mov)

Is there an easy way to disable `perfect-scrollbar` during runtime to check if I can reproduce my findings in our production Grafana?

Thanks for digging into this.

**Comment by daniellee** [comment]
@dictvm thanks for the videos. This looks similar to when Torkel showed me scrolling perf on his macbook. We tested on a different macbook today and the scrolling is fine there. So not sure what the difference is. I'll be talking to Torkel tomorrow so we can do some more testing with his macbook.

I sort of see the same thing on iOS (but I'm not sure if it is the same as cannot reproduce on macOS) and I think I might have figured out what is causing it - one line of CSS. 

I think by setting `overflow: auto`, it is turning off momentum scrolling. So it seems it is not due to perfect-scrollbar - it is this line of code: https://github.com/grafana/grafana/blob/master/public/sass/layout/_page.scss#L29 

Going to do some more testing before submitting a fix.

The other problem with nested divs not scrolling is a perfect-scrollbar problem and is not related to this.

**Comment by antiquechrono** [comment]
I've noticed that while scrolling it tends to suddenly stop when the mouse enters the legend of a graph. I then tried scrolling with the mouse over the scroll bar and it's much more smooth. I assume the rail-x and rail-y inside the legend are some kind of scroll bar elements so maybe it's getting hung up on them?

**Comment by daniellee** [comment]
I've spent most of the day working on this and almost got it working but in the end couldn't get everything working and behaving it as it should. 

The main problem is that the perfect-scrollbar turns off momentum scrolling on iOS. 

I tested not loading the perfect-scrollbar component for small screen sizes but then you lose the scroll position (when first navigating into a panel and then back to the dashboard). The CSS gets very complicated as any changes I could think of either impacted the side menu (with its absolute positioning) or the scroll position.

I did however dig up one small improvement - faster scrolling when using the mouse wheel and that will be released in 5.0.1 tomorrow.

**Comment by thenayr** [comment]
@daniellee I don't think either of the issues you've mentioned are the core of the problem presented in this issue.  I did notice the momentum scrolling is absent, but that doesn't account for the overall performance of the page.  

The issue is much more prevalent when using the built-in MacBook screen for me than it is connected to an external monitor, likely has to do with refresh rate.  

I have one dashboard with ~15 charts on it that updates every 10 seconds.  The scrolling went from silky smooth on pre 5.0 to noticeably laggy and choppy (think 60fps to closer to 15-20fps) in 5.0.  

I'll try and put together a video if you think it would help...but really I think the explanation covers the behavior adequately.

**Comment by bpbp-boop** [comment]
As I commented in https://github.com/grafana/grafana/issues/10810, perhaps replacing perfect-scrollbar with baron would fix these scrolling issues?

> Basically it seems perfect-scrollbar emulates scrolling while baron hides the native scrollbar, lets the browser handle scrolling and you can then style your own custom one.

**Comment by daniellee** [comment]
@thenayr we will keep investigating this - thanks for the gifs. Will do some performance testing with and without perfect-scrollbar. A question - are you using a mouse or the trackpad? 

(Just did a quick test on Ubuntu and not seeing any differences in FPS between 4.6 and 5.0. Also noticed that FPS is > 50 when scrolling with the mouse wheel but drops a lot more using the trackpad on my Dell XPS for both 4.6 and 5.0. But that is Ubuntu and not MacOS so don't think I can draw any conclusions from that.)

@bpbp-boop thanks - saw that already and am planning to test it out.

**Comment by thenayr** [comment]
I'm using a trackpad (MacBook built-in)

**Comment by viktorsmari** [comment]
Not sure if related, but in Grafana 5 I feel the scroll is too fast, or scrolls too much at a time, very different from the normal / expected behaviour. 

I would rather not to have some external library change how much or how fast my mouse scrolls between websites. This is especially annoying when I have a small window height, and one scroll increment is scrolling 33% of the page out of view.

**Comment by fake-name** [comment]
Please remove the current behaviour until something that's not so broken is an option. 

Having the page look ugly is significantly preferable to what we have now. Proper functionality is far more important then looking pretty.

**Comment by daniellee** [comment]
@fake-name it is not just a cosmetic feature unfortunately so removing it is not that easy.

**Comment by tagliola** [comment]
Scrolling degradation after upgrade to 5.0.2 is highly noticable on mobile as well, using Chrome 65 on Android 7.0.

**Comment by wridgers** [comment]
I'm also getting serious scroll issues with Firefox 61.

  - Scrolling is sluggish.
  - Can't scroll with click/drag until I first click the page.
  - Can't scroll with page up/down or space/shift+space without first clicking the page.
  - If you can't see the scrollbar, it's not obvious there is more content below the fold.

> we are using a js framework to handle the scrollbar as the native scrollbar takes up space and looks absolutely horrible

I don't think you should sacrifice usability for being pretty. To me, this is a good example of why you shouldn't mess with native browser functionality like scrollbars or form elements.

**Comment by Josh-G** [comment]
Scrolling is also janky on Chrome on Android. Samsung S7 has smooth scrolling in a complex dashboard on  4.x but appears to have a significant fps hit in 5 even on small dashboards.

![example](https://user-images.githubusercontent.com/487384/37620605-a5e20db4-2bb4-11e8-9413-8bc97f3946c0.gif)
v4 on left, v5 on right. Quality of gif has degraded a little but should still be clear.

**Comment by rolandjohann** [comment]
It seems that there is some JavaScript magic to customize scroll behavior. When under heavy load the site loads slow and as soon as `window.load` or `document.ready` kicks in scrolling is different - and sluggy.
I never can understand why some web apps or websites overwrite default scrolling behavior, because it always feels different and strange.
Is there an ability to turn this off?

**Comment by fake-name** [comment]
Here's a handy flow-chart for deciding of you need to override browser behaviour:

![image](https://user-images.githubusercontent.com/1401239/37858233-90696f46-2ebf-11e8-8371-57d8f1cb14fc.png)

**Comment by torkelo** [comment]
It’s not like we wanted to this, only chrome has native overlay scrollbars, and only enabled by hidden feature toggle, safari and Firefox has no such feature yet.

But we will try to change scroll component to a better one that keeps native perf and behavior.

**Comment by fake-name** [comment]
The problem isn't the *choice* of the scroll component, **It's that you're trying to override scrolling AT ALL**.

Leave the ugly native scroll bars. That's much, MUCH better then broken.

**Comment by torkelo** [comment]
Broken? Maybe in some edge cases , we had thousands of users and installs of beta nd stable before complaints came in that in some cases the perf and behavior was not ideal. We had not notified anything during development. 

But as I said we are aware that some scenarios are not working great and we will work to address that. Sorry for not spotting this during beta.

**Comment by fake-name** [comment]
Well, extremely annoying, if not broken.

And if by "edge cases", you mean "using a mouse with a scroll wheel on windows".

**Comment by torkelo** [comment]
Strange, can’t replicate any weird or slow scroll behavior on windows. What browser are you using ?

**Comment by fake-name** [comment]
All of them (chrome, vivaldi, firefox).

And the issue I have (https://github.com/grafana/grafana/issues/11267) which was marked as a dupe of this is not that it's sluggish, it's that the scrolling is in discrete, instantaneous jumps, rather then what the browser does, which is animate the scroll. Additionally, the os preferences of how much to move on each scroll tick are ignored.

**Comment by fake-name** [comment]
How scrolling everywhere else works:

![expected scrolling](https://user-images.githubusercontent.com/1401239/37860760-9e575846-2ee9-11e8-87bc-90a4659d871b.gif)

In grafana only:

![jumpy mess](https://user-images.githubusercontent.com/1401239/37860763-a774b77a-2ee9-11e8-8e8d-cddd72e9ac4f.gif)

**Comment by gjcarneiro** [comment]
So, all this just because:

> the native scrollbar takes up space and looks absolutely horrible

I disagree with "looks absolutely horrible", and I don't mind about "takes up space".  There are reasons why the native scrollbar looks and behaves the way it does.

I suggest to bring back native scrollbars by default, and enable JS scrollbars only as a preference, a visual style option, if at all.

**Comment by torkelo** [comment]
Not sure how many times we can say that we agree with you and we will address it :) It's not a quick simple fix though as the scrollbar component solves many other issues as well and simply removing it is not an option.

**Comment by elfchief** [comment]
What other issues does replacing native scrolling solve? I keep seeing that in this thread as a reason for this awfulness (and it *is* awful, at the moment), but no actual details. Every other version of grafana before 5 has seemed fine, so I'm wondering what sudden new critical issue has popped up that required overriding such a core part of the browser's functionality...

**Comment by tsia** [comment]
i did some digging when 5.0 was released and if i remember correctly even 4.x used some custom scrolling magic. but during 5.0 development it was changed 3 or 4 times.

**Comment by tsia** [comment]
And this is for the „broken“ and „edge case“ discussion
[Video](https://nofile.io/f/Hc0gd4PQZJm/12EDB69A-E1BD-46A1-B96E-8A685A3AF205.MP4)

**Comment by lfdominguez** [comment]
Using js for scrollbar break some plugins to take a full page photo of the dashboard. I think that the best way is let that browser use their own scrollbars....

**Comment by daniellee** [comment]
Closed by #11456

**Comment by vonox7** [comment]
@torkelo "Broken? Maybe in some edge cases , we had thousands of users and installs of beta nd stable before complaints came in that in some cases the perf and behavior was not ideal. We had not noticed anything during development, if we had we would not have released it with this component."

I also didn't succeed to reproduce the issue in all public dashboards from grafana I found. The big difference between them and out internal (currently mobile 100% unusable) dashboards is that we have per dashboard 12-30 charts with each 5-40 data-lines, and all data-lines have max/avg/current value plotted below or on their side. Could you please add also some bigger dashboards with a lot of queries per panel to your test suite and play.grafana.org ?

**Comment by fake-name** [comment]
@tsia - Uh, are you trying to demonstrate it's broken, or claim it's not?

My issue here was originally that it's broken *with a scroll wheel mouse*, which has exactly nothing to with that preview.

That video shows it's broken pretty well, actually, albeit in a different way. Almost every phone browser shrinks the URL bar on scroll events, to give the viewport more space. However, because the scroll is completely overridden, the URL bar and nav bar never collapse. This is a **major** issue if you're trying to use landscape view. 

TL;DR If you are overriding browser scrolling behaviour, you are **WRONG** (in 99.99% of use cases, *including this one*)..

**Comment by bwesen** [comment]
I know this is closed now and work seems to be well underway to solve it in the referenced PR discussion above. Just wanted to chime in and say I had the exact same problem that fake-name showed a gif on further above, when scrolling using the touchpad on the MacBook Pro even simple dashboards jump in gigantic steps at maybe 2-3 fps, before this release scrolling had been completely normal for years.
